extensions/arc/deployment/src/test/java/io/quarkus/arc/test/synthetic/restrictTypes/SyntheticBeanBuildItemRestrictTypesTest.java

```diff
@@ -0,0 +1,109 @@
+import io.quarkus.builder.BuildContext;
+import io.quarkus.builder.BuildStep;
+import io.quarkus.test.QuarkusUnitTest;
+
+public class SyntheticBeanBuildItemRestrictTypesTest {
+
+    @RegisterExtension
+    static final QuarkusUnitTest config = new QuarkusUnitTest()
+            .withApplicationRoot((jar) -> jar
+                    .addClasses(SyntheticBeanBuildItemRestrictTypesTest.class, FooCreator.class, FooInterface.class, Foo.class,
+                            FooSubclass.class, Charlie.class, CharlieSubclass.class, CharlieInterface.class, BarInterface.class,
+                            BazInterface.class))
+            .addBuildChainCustomizer(buildCustomizer());
+
+    static Consumer<BuildChainBuilder> buildCustomizer() {
+        return new Consumer<BuildChainBuilder>() {
+
+            @Override
+            public void accept(BuildChainBuilder builder) {
+                builder.addBuildStep(new BuildStep() {
+
+                    @Override
+                    public void execute(BuildContext context) {
+                        context.produce(SyntheticBeanBuildItem.create(FooSubclass.class)
+                                .addTypeClosure(FooSubclass.class)
+                                .removeTypes(DotName.createSimple(CharlieSubclass.class))
+                                .removeTypes(DotName.createSimple(FooSubclass.class))
+                                .removeTypes(Type.create(BazInterface.class))
+                                .scope(BuiltinScope.SINGLETON.getInfo())
+                                .unremovable()
+                                .creator(FooCreator.class)
+                                .done());
+                    }
+                }).produces(SyntheticBeanBuildItem.class).build();
+            }
+        };
+    }
+
+    @Test
+    public void testRestrictingBeanTypes() {
+        ArcContainer container = Arc.container();
+        Assertions.assertTrue(container.select(Foo.class).isResolvable());
+        Assertions.assertTrue(container.select(FooInterface.class).isResolvable());
+        Assertions.assertTrue(container.select(BarInterface.class).isResolvable());
+        Assertions.assertTrue(container.select(Charlie.class).isResolvable());
+        Assertions.assertTrue(container.select(CharlieInterface.class).isResolvable());
+
+        // CharlieSubclass, FooSubclass and BazInterface should not be registered as bean types
+        Assertions.assertFalse(container.select(CharlieSubclass.class).isResolvable());
+        Assertions.assertFalse(container.select(FooSubclass.class).isResolvable());
+        Assertions.assertFalse(container.select(BazInterface.class).isResolvable());
+    }
+
+    @Vetoed
+    public static class FooSubclass extends Foo implements FooInterface {
+    }
+
+    @Vetoed
+    public static class Foo extends Charlie implements BazInterface {
+    }
+
+    @Vetoed
+    public static class CharlieSubclass extends Charlie {
```

Comment:
This is not in the type closure of `FooSubclass`. Is that intended? Just wondering.

Reply:
Yes, this was intentional.

Reply:
Gotcha, thanks!